### PR TITLE
Core/Spells: Added helper to get clean and unmitigated damage for pro…

### LINKED ISSUE #6900: Rogue "Blade Flurry" Talent - Secondary Target Damage Bugged
#### WHAT CLIENT DO YOU PLAY ON?

- [ ] enGB
- [x] enUS
- [ ] other (specify)

##### FACTION

- [x] Alliance
- [x] Horde

##### CONTENT PHASE:

- [x] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

##### CURRENT BEHAVIOUR:

Original report: https://github.com/chromiecraft/chromiecraft/issues/1167

Blade Flurrys(Rogue Talent https://wotlk-twinhead.twinstar.cz/?spell=13877)  behavior is not correct, it calculates the Armor reduction of the second Target two times. for Example if i hit 2 identical monsters, the main target gets hit for 500 while the second target gets hit for 250, getting reduced by armor again. 

Tester: The second target is definitely taking consistently less damage than the first. Checking combat logs showed it was ~22% less damage on average on target dummies with identical armour. Can't say if it's armour reduction of the second target being counted twice as the reporter suggests, but it sounds pretty plausible.

WoWScrnShot_071221_213747
WoWScrnShot_071221_213751
WoWScrnShot_071221_213755
WoWScrnShot_071221_213813

Note Blade Flurry was worked on in https://github.com/azerothcore/azerothcore-wotlk/pull/939/files

@jamadaha adds:
Followed the steps on my local instance (AzerothCore 9b3993b). There does seem to be less damage done to the second target.
image
image
image
image

##### EXPECTED BLIZZLIKE BEHAVIOUR:

the second Target should receive the same damage as the main target if their armor values are the same, not getting reduced again. here is a old Youtube Video from Wotlk of a Rogue hitting 2 Targets with Bladeflurry while both receive around the same dmg https://www.youtube.com/watch?v=ats4GrMUgVc.

Long discussion on Blade Flurry in the Wrath era here: https://www.mmo-champion.com/threads/675288-How-does-Blade-Flurry-work - though a lot of confusion as people talk past and misunderstand each other.

"If an ability hits the first mob for 1000 before damage mitigation, the second mob will also be hit for 1000 damage BEFORE damage mitigation." - so identical damage mitigation should mean identical damage.

"The original damage is mirrored, but the actual damged will vary on the state of the mob. EG, I hit the first mob for 10k, it has no armor so hits for the full 10k. That 10k hit is then sent to the 2nd mob however it has 50% DR from armor so it only gets hit for 5k."

##### SOURCE:

https://www.youtube.com/watch?v=ats4GrMUgVc
https://www.mmo-champion.com/threads/675288-How-does-Blade-Flurry-work
https://tbc.wowhead.com/spell=13877/blade-flurry

##### STEPS TO REPRODUCE THE PROBLEM:

1.  get on a atleast level 30 Rogue with the Talent Blade Flurry specced
2.  pull 2 mobs with near the same Armor Values and hit them while Blade Flurry is activated
3.  notice the second target gets significantly less damage than the first

##### EXTRA NOTES:

multiple more Videos for evidence! also its the way it works in Classic WoW at the moment, while the Spell didnt got changed in Wotlk.

##### AC HASH/COMMIT:

https://github.com/chromiecraft/azerothcore-wotlk/commit/7c21d567d59153b37978756a5bf02daf6998e665

##### OPERATING SYSTEM:

Ubuntu 20.04

##### MODULES:

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-CarbonCopy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-LevelUpReward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-Zonecheck](https://github.com/55Honey/acore_Zonecheck)

##### OTHER CUSTOMIZATIONS:

None.

##### SERVER:

ChromieCraft

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/99900433-rogue-blade-flurry-talent-secondary-target-damage-bugged?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #7636: Warrior "Sweeping Strikes" Ability - Secondary Target Damage Bugged
### What client do you play on?

enUS

### Faction

- [X] Alliance
- [X] Horde

### Content Phase:

- [X] Generic
- [ ] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

### Current Behaviour
Original report: https://github.com/chromiecraft/chromiecraft/issues/1580

When the buff "Sweeping Strikes" is up on the warrior (Arms talent ability) the five attacks that strike an additional opponent deal different damage (always lower) than the main attack they copy. I found the same exact bug a couple years ago on another server named TrueWow and helped to fix it. It behave like the armor of the "additional nearby opponent" is further reducing the damage of the sweeping strike hit -which is not correct because the damage of the hit that will be "copied" is already reduced by the first opponent's armor.

Tester: confirmed. Tried this for a bit and the secondary target damage was always less than the primary, and always by the same percentage ratio. It looks like the secondary target's armour is being counted twice, as was the case for the similar issue involving the rogue Blade Flurry skill - see https://github.com/azerothcore/azerothcore-wotlk/issues/6900

WoWScrnShot_083021_211256
WoWScrnShot_083021_211231
WoWScrnShot_083021_211233
WoWScrnShot_083021_211236
WoWScrnShot_083021_211253

### Expected Blizzlike Behaviour

When the buff "Sweeping Strikes" is up on the warrior (Arms talent ability) the five copied attacks should be of the exact amount of damage of the original hit.

"It tries to copy your direct damage attacks to a secondary target, exactly like the rogue Blade Flurry talent, except SS is limited to only 5 attacks. ... Basically, SS will copy the instant damage of a skill but not the effect. SS + hamstring will do 50ish damage on both targets but only hamstring the primary target, for instance."

"Sweeping Strikes only duplicates the damage from your hits." - https://www.mmo-champion.com/threads/602307-Sweeping-Strikes-and-Mortal-Strike?p=5521686&viewfull=1#post5521686

### Source

https://hilbert.wordpress.com/sweeping-strikes-guide (Classic, but mechanics are the same.)
https://wowpedia.fandom.com/wiki/Sweeping_Strikes?oldid=2378895

### Steps to reproduce the problem

1) Activate Sweeping Strikes with an Arms Warrior
2) Need two-three enemies to hit in order to make Sweeping Strikes hit a nearby opponent
3) Observe the difference in damage between the original hit and the sweeping strike hit copied.

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/d20f315bd5370761f226460964ae506a97f372c0

### Operating system

Ubuntu 20.04

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-custom-corldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by UltraNix**
…c system.

Fixed #6900

## Changes Proposed:
-  Added helper to get clean and unmitigated damage for proc system
-  Fixed damage of Blade Flurry.
-  Fixed damage of Sweeping Strikes.

**Eluna buildifx handled [here](https://github.com/ElunaLuaEngine/Eluna/pull/381)**

## Issues Addressed:

- Closes #6900
- Closes #7636
- Closes https://github.com/chromiecraft/chromiecraft/issues/1167
- Closes https://github.com/chromiecraft/chromiecraft/issues/1580

## Tests Performed:

- Tested ingame.

## How to Test the Changes:

1. get on a atleast level 30 Rogue with the Talent Blade Flurry specced
2. pull 2 mobs with near the same Armor Values and hit them while Blade Flurry is activated
3. notice the second target gets significantly less damage than the first

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -771,12 +771,21 @@ uint32 Player::EnvironmentalDamage(EnviromentalDamage type, uint32 damage)
     // Absorb, resist some environmental damage type
     uint32 absorb = 0;
     uint32 resist = 0;
```
Comment (Helias) [inline src/server/game/Entities/Player/Player.cpp:773]: this is unused probably, so the build fail

File: src/server/game/Spells/SpellEffects.cpp [hunk 2]
```diff
@@ -294,11 +294,21 @@ void Spell::EffectEnvironmentalDMG(SpellEffIndex /*effIndex*/)
     uint32 absorb = 0;
     uint32 resist = 0;
```
Comment (Helias) [inline src/server/game/Spells/SpellEffects.cpp:295]: is it used? I see that it's declared below

File: src/server/game/Entities/Player/Player.cpp (hunk 1 above)
Comment (UltraNix) [reply src/server/game/Entities/Player/Player.cpp:773]:
```
WorldPacket data(SMSG_ENVIRONMENTALDAMAGELOG, (21));
    data << GetGUID();
    data << uint8(type != DAMAGE_FALL_TO_VOID ? type : DAMAGE_FALL);
    data << uint32(damage);
    data << uint32(absorb);
    data << uint32(resist);
    SendMessageToSet(&data, true);
```

**Comment by FrancescoBorzi** [comment]
@UltraNix please fix the merge conflicts

**Comment by pklloveyou** [comment]
image
Tested?

**Comment by FrancescoBorzi** [comment]
@pklloveyou can you please describe what's happening? we don't understand Chinese :(

**Comment by pklloveyou** [comment]
The white words on the red line are skill damage and ordinary damage. Is this damage correct? Did you pass the test?

**Comment by UltraNix** [comment]
What mobs did you test?

**Comment by UltraNix** [comment]
ad

**Comment by pklloveyou** [comment]
https://github.com/azerothcore/azerothcore-wotlk/issues/7279 
Does it have anything to do with here? I merged too many PR's to be tested at one time

**Comment by UltraNix** [comment]
Rebase to newest AC version and then test that PR again.

**Comment by UltraNix** [comment]
Yep - bad code.

**Comment by UltraNix** [comment]
@pklloveyou Should be okay now.

**Comment by pklloveyou** [comment]
Is it necessary to do so -> https://github.com/azerothcore/azerothcore-wotlk/pull/7281

**Comment by UltraNix** [comment]
That crashfix is related to that PR.

**Comment by FrancescoBorzi** [comment]
@UltraNix build failing

**Comment by FrancescoBorzi** [comment]
@azerothcore/developers @azerothcore/testers @pklloveyou please check this

**Comment by FrancescoBorzi** [comment]
any updates on this?

**Comment by UltraNix** [comment]
Sweeping Strikes part needs tests.

**Comment by FrancescoBorzi** [comment]
@UltraNix there are conflicts

**Comment by Helias** [comment]
build failing
```
In file included from /home/runner/work/azerothcore-wotlk/azerothcore-wotlk/modules/mod-eluna-lua-engine/LuaEngine/LuaFunctions.cpp:23:
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/modules/mod-eluna-lua-engine/LuaEngine/UnitMethods.h:2818:50: fatal error: too many arguments to function call, expected at most 2, have 7
            unit->CalcAbsorbResist(unit, target, schoolmask, SPELL_DIRECT_DAMAGE, damage, &absorb, &resist);
            ~~~~~~~~~~~~~~~~~~~~~~               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/game/Entities/Unit/Unit.h:2309:5: note: 'CalcAbsorbResist' declared here
    static void CalcAbsorbResist(DamageInfo& dmgInfo, bool Splited = false);
    ^
1 error generated.
```

**Comment by UltraNix** [comment]
@Helias https://github.com/ElunaLuaEngine/Eluna/pull/381

**Comment by FrancescoBorzi** [comment]
> ElunaLuaEngine/Eluna#381

can we have that merged please @UltraNix @Rochet2 ? so AC build will be green again

**Comment by UltraNix** [comment]
Okay.

**Comment by Rochet2** [comment]
done

**Comment by UltraNix** [comment]
As I tested it, let's just merge it. Eluna is not building without this PR.

**Comment by IntelligentQuantum** [comment]
Tested

**Review by IntelligentQuantum — APPROVED** [review]

**Review by Nefertumm — APPROVED** [review]
Seems ok to me, I'm re-running the actions in order to merge this tomorrow :)

**Comment by Nyeriah** [comment]
Build is failing

```/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/modules/mod-eluna-lua-engine/LuaEngine/UnitMethods.h:2818:50: fatal error: too many arguments to function call, expected at most 2, have 7
            unit->CalcAbsorbResist(unit, target, schoolmask, SPELL_DIRECT_DAMAGE, damage, &absorb, &resist);
```

**Comment by Rochet2** [comment]
> Build is failing, so merging not a possibility still
> 
> ```
> /home/runner/work/azerothcore-wotlk/azerothcore-wotlk/modules/mod-eluna-lua-engine/LuaEngine/UnitMethods.h:2818:50: fatal error: too many arguments to function call, expected at most 2, have 7
>             unit->CalcAbsorbResist(unit, target, schoolmask, SPELL_DIRECT_DAMAGE, damage, &absorb, &resist);
> ```

The CI looks like its not using latest Eluna, but rather its maybe using the submodule commit hash which is not probably pointing at the latest version.

**Comment by UltraNix** [comment]
@Rochet2 is right - there is something wrong with CI.

**Pull request merged**